### PR TITLE
[cli] Allow projectSettings override in vercel.json

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
This PR adds the required changes to the CLI to support overridable `vercel.json` `projectSettings` values. This PR is linked with changes on the API side too.

- The core changes are in `packages/cli/src/commands/deploy/index.ts`
- An unused code path was removed from `packages/client/src/create-deployment.ts`, `packages/cli/src/util/deploy/process-deployment.ts` and `packages/cli/src/util/index.ts`.

This will only work for *existing* deployment (not *new* ones). This is implemented by checking the `status` of the deployment (is it linked or not). An error is thrown if the user attempts to send overrided settings to a new project. 

A warning is outputted when overrided settings are discovered and shares what settings are being overridden.

#### Tests

- [x] The code changed/added as part of this PR has been covered with tests
- [x] All tests pass locally with `yarn test-unit`

#### Code Review

- [x] This PR has a concise title and thorough description useful to a reviewer
- [ ] Issue from task tracker has a link to this PR

### CODE REVIEW

File: packages/cli/src/util/deploy/create-deploy.ts [hunk 1]
```diff
@@ -19,7 +19,7 @@ export default async function createDeploy(
   cwd?: string
 ): Promise<any | DeploymentError> {
   try {
-    return await now.create(paths, createArgs, org, isSettingUpProject, cwd);
+    return now.create(paths, createArgs, org, isSettingUpProject, cwd);
```
Comment (TooTallNate) [inline packages/cli/src/util/deploy/create-deploy.ts]: We still want `await` here otherwise a thrown error won't be caught in the try/catch here.

Comment (Ethan-Arrowood) [reply packages/cli/src/util/deploy/create-deploy.ts]: i feel like the async/await is the wrong pattern here, but i don't think this is a high volume promise call so there is little point trying to optimize. If it makes the error handling easier than no need to change either 😄

File: packages/cli/src/commands/deploy/index.ts [hunk 2]
```diff
@@ -460,7 +460,10 @@ export default async (client: Client) => {
 
     if (!localConfig.builds || localConfig.builds.length === 0) {
       // Only add projectSettings for zero config deployments
-      createArgs.projectSettings = { sourceFilesOutsideRootDirectory };
+      createArgs.projectSettings = {
+        ...localConfig.projectSettings,
```
Comment (Ethan-Arrowood) [inline packages/cli/src/commands/deploy/index.ts:472]: this is the line that applies the projectSettings to the deployment

File: packages/cli/src/commands/deploy/index.ts [hunk 3]
```diff
@@ -321,7 +321,7 @@ export default async (client: Client) => {
       `${prependEmoji(
         `The ${code('name')} property in ${highlight(
           localConfig[fileNameSymbol]!
-        )} is deprecated (https://vercel.link/name-prop)`,
+        )} is deprecated (https://vercel.link/name-prop). Use the settings UI or projectSettings.projectName instead.`,
```
Comment (styfle) [inline packages/cli/src/commands/deploy/index.ts]: I don't think we want to encourage projectName either

File: packages/client/src/create-deployment.ts [hunk 4]
```diff
@@ -76,29 +72,6 @@ export default function buildCreateDeployment() {
...
-      // If the user did not provide a config file, use the one in the root directory.
-      const relativePaths = fileList.map(f => relative(cwd, f));
-      const hasVercelConfig = relativePaths.includes('vercel.json');
-      const hasNowConfig = relativePaths.includes('now.json');
-
-      if (hasVercelConfig) {
-        if (hasNowConfig) {
-          throw new DeploymentError({
-            code: 'conflicting_config',
-            message:
```
Comment (styfle) [inline packages/client/src/create-deployment.ts:90]: Why was this removed?

Comment (Ethan-Arrowood) [reply packages/client/src/create-deployment.ts:90]: `configPath` and `nowConfig` are never used throughout the function.

Comment (Ethan-Arrowood) [reply packages/client/src/create-deployment.ts:90]: so this whole block is useless; and thus the `nowConfig` function parameter is similarly useless.

Comment (styfle) [reply packages/client/src/create-deployment.ts:90]: Strange because there is a test for this error message here

https://github.com/vercel/vercel/blob/b2bfae6e2e15237103607ab0f783617a0af1e45e/packages/cli/test/integration.js#L3210-L3229

Comment (Ethan-Arrowood) [reply packages/client/src/create-deployment.ts:90]: Yes because the if condition will still run when the config is omitted, but since that config is never used in this block I think the error message is pointless too. I will remove that test as well

File: packages/cli/src/commands/deploy/index.ts [hunk 5]
```diff
@@ -153,6 +153,27 @@ export default async (client: Client) => {
...
+        'framework',
+        'name',
+        'nodeVersion',
+        'commandForIgnoringBuildStep',
+        'rootDirectory',
+        'sourceFilesOutsideRootDirectory',
+      ];
+      for (const key of Object.keys(localConfig.projectSettings)) {
+        if (!allowedProperties.includes(key)) {
+          output.error(`Disallowed property ${key} found in "projectSettings"`);
```
Comment (styfle) [inline packages/cli/src/commands/deploy/index.ts]:
```suggestion
          output.error(`Unknown property ${key} found in "projectSettings"`);
```

We should consider this unknown just like if there was any other unknown property.

What happens when you do `{ projectSettings: { "foo": "bar" } }`?

Comment (Ethan-Arrowood) [reply packages/cli/src/commands/deploy/index.ts]: I think you'll eventually hit an API error when this object is validated by the API using the schema

Comment (styfle) [reply packages/cli/src/commands/deploy/index.ts]: Lets remove this and validate in the API instead of the CLI

File: packages/client/src/create-deployment.ts (hunk 4 above)
Comment (styfle) [reply packages/client/src/create-deployment.ts:90]: @Ethan-Arrowood We need to keep the test. We must throw when both now.json exist and vercel.json exist. I suspect the error might be thrown elsewhere too which is why the test still passed.

Comment (Ethan-Arrowood) [reply packages/client/src/create-deployment.ts:90]: Upon review, this error is thrown from elsewhere in the code base. The test passes with no issue.

File: packages/cli/src/commands/deploy/index.ts (hunk 5 above)
Comment (Ethan-Arrowood) [reply packages/cli/src/commands/deploy/index.ts]: So the create-deployment api route will need to use a different schema than update-project. Okay ✅

File: packages/cli/test/integration.js [hunk 6]
```diff
@@ -3580,3 +3580,76 @@ test('[vc link] should support the `--project` flag', async t => {
...
+      stdout: deployment.stdout,
+    })
+  );
+
+  const page = await fetch(deployment.stdout);
+  const text = await page.text();
+  t.is(text, `Hello, World\n`);
+});
+
+test('invalid vercel.json projectSettings should result in a deployment error', async t => {
```
Comment (Ethan-Arrowood) [inline packages/cli/test/integration.js]: This test won't pass until API is updated. This gives me good reason to focus on merging API changes first and then we can wrap up these CLI changes.

File: packages/build-utils/src/types.ts [hunk 7]
```diff
@@ -320,17 +320,29 @@ export interface BuilderFunctions {
...
   rootDirectory?: string | null;
-  createdAt?: number;
-  autoExposeSystemEnvs?: boolean;
+  serverlessFunctionRegion?: string | null;
   sourceFilesOutsideRootDirectory?: boolean;
-  directoryListing?: boolean;
-  gitForkProtection?: boolean;
+  ssoProtection?: {
+    deploymentType: 'all' | 'preview';
+  };
```
Comment (styfle) [inline packages/build-utils/src/types.ts]: Do we need two types now? One to represent overridable project settings and one to represent all setting on the project?

File: packages/cli/src/commands/deploy/index.ts (hunk 2 above)
Comment (styfle) [reply packages/cli/src/commands/deploy/index.ts:472]: You might also need to add this to the `missing_project_settings` error handling.

You can test this by making two deployments to the same project. The first one will return early and enter the `missing_project_settings` branch and ask the use to edit the settings if they wish to override. If the `vercel.json` has settings then those should settings automatically populate instead of asking the user again. Then the second deployment should not hit the `missing_project_settings` branch and accept the deployment on the first api call.

File: packages/cli/test/integration.js [hunk 8]
```diff
@@ -3675,3 +3675,79 @@ test('[vc link] should support the `--project` flag', async t => {
...
+  // Pass a disallowed command
+  vercelConfig['projectSettings'] = {
+    autoExposeSystemEnvs: true,
+  };
+
+  await writeFile(vercelJsonPath, JSON.stringify(vercelConfig));
+
+  await testDeployment();
+
+  await writeFile(vercelJsonPath, vercelJsonOriginal);
```
Comment (styfle) [inline packages/cli/test/integration.js]: We also need a test that creates a new deployment without custom project settings, then modifies the project settings for a second deployment via vercel.json

File: packages/cli/test/integration.js [hunk 9]
```diff
@@ -3675,3 +3675,79 @@ test('[vc link] should support the `--project` flag', async t => {
...
+  async function testDeployment() {
+    try {
+      await execa(binaryPath, [
+        directory,
+        ...defaultArgs,
+        '--public',
+        '--confirm',
+      ]);
+      t.fail('Deployment should fail');
+    } catch ({ exitCode, stderr, stdout }) {
```
Comment (styfle) [inline packages/cli/test/integration.js]: You can do this without try/catch by passing `reject: false` option to execa (see other tests). That will ensure the test fails if the exit code was successful.

File: packages/cli/test/integration.js [hunk 10]
```diff
@@ -3675,3 +3675,79 @@ test('[vc link] should support the `--project` flag', async t => {
...
+      );
+    }
+  }
+
+  const vercelJsonPath = path.join(directory, 'vercel.json');
+  const vercelJsonOriginal = await readFile(vercelJsonPath, 'utf8');
+  const vercelConfig = JSON.parse(vercelJsonOriginal);
+
+  // Pass a command that does not exist
+  vercelConfig['projectSettings'] = {
```
Comment (styfle) [inline packages/cli/test/integration.js]:
```suggestion
  vercelConfig.projectSettings = {
```

File: packages/cli/test/integration.js [hunk 11]
```diff
@@ -3675,3 +3675,79 @@ test('[vc link] should support the `--project` flag', async t => {
...
+    command_does_not_exist: 'echo "foo bar"',
+  };
+
+  await writeFile(vercelJsonPath, JSON.stringify(vercelConfig));
+
+  await testDeployment();
+
+  // Pass a disallowed command
+  vercelConfig['projectSettings'] = {
```
Comment (styfle) [inline packages/cli/test/integration.js]:
```suggestion
  vercelConfig.projectSettings = {
```

File: packages/cli/test/integration.js [hunk 12]
```diff
@@ -3675,3 +3675,79 @@ test('[vc link] should support the `--project` flag', async t => {
...
 });
+
+test('vercel.json projectSettings should be applied to deployment', async t => {
+  const directory = path.join(__dirname, 'fixtures/unit/vercel-json-overrides');
+
+  const deployment = await execa(binaryPath, [
+    directory,
+    ...defaultArgs,
+    '--public',
+    '--confirm',
```
Comment (styfle) [inline packages/cli/test/integration.js]: We need one more test that doesn't use `--confirm`.

In that case, I think it should utilize vercel.json to pre-populate answers to the questions and only ask the questions that vercel.json didn't answer.

This would match how it works for the https://vercel.com/new page when importing a new git repo.

File: packages/cli/src/commands/deploy/index.ts (hunk 2 above)
Comment (Ethan-Arrowood) [reply packages/cli/src/commands/deploy/index.ts:472]: Wouldn't the settings from `vercel.json` be applied on the first deployment?

File: packages/cli/test/integration.js [hunk 13]
```diff
@@ -3675,3 +3675,140 @@ test('[vc link] should support the `--project` flag', async t => {
...
+
+  const page = await fetch(deployment.stdout);
+  const text = await page.text();
+  t.is(text, `Hello, World\n`);
+});
+
+test('invalid vercel.json projectSettings should result in a deployment error', async t => {
+  const directory = path.join(
+    __dirname,
+    'fixtures/unit/vercel-json-overrides-error'
```
Comment (styfle) [inline packages/cli/test/integration.js]: All of these tests should be moved from `fixtures/unit` to `fixtures/integration`

File: packages/cli/test/integration.js [hunk 14]
```diff
@@ -3675,3 +3675,140 @@ test('[vc link] should support the `--project` flag', async t => {
...
+      ...defaultArgs,
+      '--public',
+    ]);
+
+    return deployment;
+  }
+
+  // Deploy without projectSettings
+  await writeFile(vercelJsonPath, JSON.stringify({}));
+  assertDeployment(await deploy(), 0);
```
Comment (styfle) [inline packages/cli/test/integration.js]: You also need to visit the deployment to see if the output directory was assigned properly. It should fetch the root index file (you could use index.txt instead of index.html to make this a simple content comparison)

File: packages/cli/src/commands/deploy/index.ts (hunk 2 above)
Comment (styfle) [reply packages/cli/src/commands/deploy/index.ts:472]: It depends if the new project should be created with the settings from vercel.json

File: packages/cli/test/fixtures/unit/vercel-json-overrides-error/package.json [hunk 15]
```diff
@@ -0,0 +1,3 @@
+{
+  "name": "vercel-json-overrides-error"
+}
```
Comment (styfle) [inline packages/cli/test/fixtures/unit/vercel-json-overrides-error/package.json]: Why is this one an error? I think you can probably reuse the same test fixture for the successful test and error test.

File: packages/cli/test/integration.js [hunk 16]
```diff
@@ -3675,3 +3675,140 @@ test('[vc link] should support the `--project` flag', async t => {
...
+      })
+    );
+  }
+
+  async function deploy() {
+    const deployment = await execa(binaryPath, [
+      directory,
+      ...defaultArgs,
+      '--public',
+    ]);
```
Comment (styfle) [inline packages/cli/test/integration.js]: We also need to test a single setting such as 

```
{
  projectSettings: {
    outputDirectory: "foo"
  }
}
```

Make sure it overrides one but allows defaults of the others

File: packages/cli/src/commands/deploy/index.ts [hunk 17]
```diff
@@ -195,6 +195,16 @@ export default async (client: Client) => {
   let sourceFilesOutsideRootDirectory: boolean | undefined = true;
 
   if (status === 'not_linked') {
+    if (
+      localConfig &&
+      Object.prototype.hasOwnProperty.call(localConfig, 'projectSettings')
+    ) {
```
Comment (styfle) [inline packages/cli/src/commands/deploy/index.ts]:
```suggestion
    if (localConfig?.projectSettings) {
```

File: packages/cli/src/commands/deploy/index.ts (hunk 2 above)
Comment (styfle) [reply packages/cli/src/commands/deploy/index.ts:472]: Now that we have an error above for new projects, I think this should be fine.

File: packages/cli/src/commands/deploy/index.ts (hunk 17 above)
Comment (Ethan-Arrowood) [reply packages/cli/src/commands/deploy/index.ts]: Lol i forget that this syntax exists sometimes 😅

**Review by styfle — APPROVED** [review]
Thanks!

File: packages/cli/src/commands/deploy/index.ts [hunk 18]
```diff
@@ -195,6 +195,13 @@ export default async (client: Client) => {
   let sourceFilesOutsideRootDirectory: boolean | undefined = true;
 
   if (status === 'not_linked') {
+    if (localConfig?.projectSettings) {
+      output.error(
+        'New projects cannot contain `projectSettings` in vercel.json'
```
Comment (TooTallNate) [inline packages/cli/src/commands/deploy/index.ts]: This error message is a bit confusing. Why can't they? Perhaps we should have a "More info" link that explains why this is the case and what the user should do about it.

Comment (Ethan-Arrowood) [reply packages/cli/src/commands/deploy/index.ts]: Ill add a link to the incoming projectSettings section and then expand on the reasoning there

**Review by TooTallNate — APPROVED** [review]

**Review by styfle — APPROVED** [review]

**Review by TooTallNate — APPROVED** [review]
